src/bson_compat.rs

```diff
@@ -1,31 +1,131 @@
-    fn append_ref<'a>(
+    fn append_err(&mut self, key: impl AsRef<str>, value: impl Into<RawBson>) -> RawResult<()> {
+        self.append(key, value.into())
+    }
+
+    fn append_ref_err<'a>(
         &mut self,
         key: impl AsRef<str>,
         value: impl Into<crate::bson::raw::RawBsonRef<'a>>,
-    ) {
+    ) -> RawResult<()> {
         self.append(key, value)
     }
 }
 
+#[cfg(not(feature = "bson-3"))]
+impl RawDocumentBufExt for crate::bson::RawDocumentBuf {
+    fn append_err(&mut self, key: impl AsRef<str>, value: impl Into<RawBson>) -> RawResult<()> {
+        self.append(key, value);
+        Ok(())
+    }
+
+    fn append_ref_err<'a>(
+        &mut self,
+        key: impl AsRef<str>,
+        value: impl Into<crate::bson::raw::RawBsonRef<'a>>,
+    ) -> RawResult<()> {
+        self.append_ref(key, value);
+        Ok(())
+    }
+
+    fn decode_from_bytes(data: Vec<u8>) -> RawResult<Self> {
+        Self::from_bytes(data)
+    }
+}
+
+pub(crate) trait RawArrayBufExt: Sized {
+    #[allow(dead_code)]
+    fn from_iter_err<V: Into<RawBson>, I: IntoIterator<Item = V>>(iter: I) -> RawResult<Self>;
+
+    fn push_err(&mut self, value: impl Into<RawBson>) -> RawResult<()>;
+}
+
 #[cfg(feature = "bson-3")]
-pub(crate) use crate::bson::error::Result as RawResult;
+impl RawArrayBufExt for crate::bson::RawArrayBuf {
+    fn from_iter_err<V: Into<RawBson>, I: IntoIterator<Item = V>>(iter: I) -> RawResult<Self> {
+        Self::from_iter(iter.into_iter().map(|v| v.into()))
+    }
+
+    fn push_err(&mut self, value: impl Into<RawBson>) -> RawResult<()> {
+        self.push(value.into())
+    }
+}
 
 #[cfg(not(feature = "bson-3"))]
-pub(crate) use crate::bson::raw::Result as RawResult;
+impl RawArrayBufExt for crate::bson::RawArrayBuf {
+    fn from_iter_err<V: Into<RawBson>, I: IntoIterator<Item = V>>(iter: I) -> RawResult<Self> {
+        Ok(Self::from_iter(iter))
+    }
 
-#[cfg(feature = "bson-3")]
-pub(crate) use crate::bson::error::Error as RawError;
+    fn push_err(&mut self, value: impl Into<RawBson>) -> RawResult<()> {
+        Ok(self.push(value))
+    }
+}
 
 #[cfg(not(feature = "bson-3"))]
-pub(crate) use crate::bson::raw::Error as RawError;
+pub(crate) trait RawDocumentExt {
+    fn decode_from_bytes<D: AsRef<[u8]> + ?Sized>(data: &D) -> RawResult<&Self>;
+}
+
+#[cfg(not(feature = "bson-3"))]
+impl RawDocumentExt for crate::bson::RawDocument {
+    fn decode_from_bytes<D: AsRef<[u8]> + ?Sized>(data: &D) -> RawResult<&Self> {
+        Self::from_bytes(data)
+    }
+}
+
+#[cfg(not(feature = "bson-3"))]
+#[allow(dead_code)]
+pub(crate) trait DocumentExt {
+    fn encode_to_vec(&self) -> crate::bson::ser::Result<Vec<u8>>;
+}
+
+#[cfg(not(feature = "bson-3"))]
+impl DocumentExt for crate::bson::Document {
+    fn encode_to_vec(&self) -> crate::bson::ser::Result<Vec<u8>> {
+        let mut out = vec![];
+        self.to_writer(&mut out)?;
+        Ok(out)
+    }
+}
+
+macro_rules! use_either {
```

Comment:
I know this is kind of a frivolous use of. a macro but I got _real_ tired of writing these out by hand and I find the tabular format a lot easier to read.  I can revert if you feel strongly :)